pilot/pkg/serviceregistry/serviceentry/controller.go

```diff
@@ -512,41 +521,61 @@ func (s *Controller) WorkloadInstanceHandler(wi *model.WorkloadInstance, event m
-		instance := convertWorkloadInstanceToServiceInstance(wi, services, se)
-		instances = append(instances, instance...)
-		if addressToDelete != "" {
-			for _, i := range instance {
-				di := i.DeepCopy()
-				di.Endpoint.Address = addressToDelete
-				instancesDeleted = append(instancesDeleted, di)
+		currInstance := convertWorkloadInstanceToServiceInstance(wi, services, se)
+
+		// We chech if the wi is still a subset of se. This would cover Case 1 and Case 2 from above.
+		if labels.Instance(se.WorkloadSelector.Labels).SubsetOf(wi.Endpoint.Labels) {
+			// If the workload instance still matches. We take care of the possible events.
+			instances = append(instances, currInstance...)
+			if addressToDelete != "" {
+				for _, i := range currInstance {
+					di := i.DeepCopy()
+					di.Endpoint.Address = addressToDelete
+					instancesDeleted = append(instancesDeleted, di)
+				}
+				s.serviceInstances.deleteServiceEntryInstances(seNamespacedName, key)
+			} else if event == model.EventDelete {
+				s.serviceInstances.deleteServiceEntryInstances(seNamespacedName, key)
+			} else {
+				s.serviceInstances.updateServiceEntryInstancesPerConfig(seNamespacedName, key, currInstance)
 			}
-			s.serviceInstances.deleteServiceEntryInstances(seNamespacedName, key)
-		} else if event == model.EventDelete {
-			s.serviceInstances.deleteServiceEntryInstances(seNamespacedName, key)
-		} else {
-			s.serviceInstances.updateServiceEntryInstancesPerConfig(seNamespacedName, key, instance)
-		}
-		// If serviceentry's resolution is DNS, make a full push
-		// TODO: maybe cds?
-		if (se.Resolution == networking.ServiceEntry_DNS || se.Resolution == networking.ServiceEntry_DNS_ROUND_ROBIN) &&
-			se.WorkloadSelector != nil {
-
-			fullPush = true
-			for _, inst := range instance {
-				configsUpdated[model.ConfigKey{
-					Kind:      kind.ServiceEntry,
-					Name:      string(inst.Service.Hostname),
-					Namespace: cfg.Namespace,
-				}] = struct{}{}
+			// If serviceentry's resolution is DNS, make a full push
+			// TODO: maybe cds?
+			if (se.Resolution == networking.ServiceEntry_DNS || se.Resolution == networking.ServiceEntry_DNS_ROUND_ROBIN) &&
+				se.WorkloadSelector != nil {
+
+				fullPush = true
+				for _, inst := range currInstance {
+					configsUpdated[model.ConfigKey{
+						Kind:      kind.ServiceEntry,
+						Name:      string(inst.Service.Hostname),
+						Namespace: cfg.Namespace,
+					}] = struct{}{}
+				}
 			}
+		} else if labels.Instance(se.WorkloadSelector.Labels).SubsetOf(oldWi.Endpoint.Labels) {
+			// If we're here. It means that the labels changed and the new labels don't match the SE anymore (Case 3 from above).
+			// But that doesn't mean that the old labels did match. We only care about the cases where the oldWi matched and the new one does not.
+			oldInstance := convertWorkloadInstanceToServiceInstance(oldWi, services, se)
+			unSelected := workloadInstanceDiff(oldInstance, currInstance)
```

Comment:
Do we need compare here? 
>If we're here. It means that the labels changed and the new labels don't match the SE anymore

So can we remove the old instances and add the new ones.

Reply:
You're right. We don't need it anymore. 

Also, we don't need to add the new instance here since this case only means that the WI doesn't match the SE anymore. So just deleting here is enough.